File: backend/app/services/mongodb_service.py
from pymongo import MongoClient
from pymongo.server_api import ServerApi
from bson.objectid import ObjectId
from datetime import datetime
import os
from typing import Dict, List, Any, Optional

class MongoDBService:
    def __init__(self):
        uri = os.getenv("MONGODB_URI")
        db_name = os.getenv("MONGODB_DATABASE")
        
        # Create a new client and connect to the server
        self.client = MongoClient(uri, server_api=ServerApi('1'))
        self.db = self.client[db_name]
        
        # Collections
        self.cameras = self.db["cameras"]
        self.detection_events = self.db["detection_events"]
        self.detection_zones = self.db["detection_zones"]
        self.device_info = self.db["device_info"]
        self.app_state = self.db["app_state"]
        
    def test_connection(self):
        """Test the MongoDB connection"""
        try:
            self.client.admin.command('ping')
            return True
        except Exception as e:
            print(f"MongoDB connection error: {e}")
            return False
    
    # Camera operations
    def get_cameras(self, user_id: str) -> List[Dict]:
        """Get all cameras for a user"""
        return list(self.cameras.find({"userId": user_id}))
    
    def add_camera(self, camera_data: Dict) -> str:
        """Add a new camera"""
        camera_data["createdAt"] = datetime.now()
        result = self.cameras.insert_one(camera_data)
        return str(result.inserted_id)
    
    def update_camera(self, camera_id: str, update_data: Dict) -> bool:
        """Update camera information"""
        result = self.cameras.update_one(
            {"_id": ObjectId(camera_id)},
            {"$set": update_data}
        )
        return result.modified_count > 0
    
    def delete_camera(self, camera_id: str) -> bool:
        """Delete a camera"""
        result = self.cameras.delete_one({"_id": ObjectId(camera_id)})
        return result.deleted_count > 0
    
    # Detection events operations
    def add_detection_event(self, event_data: Dict) -> str:
        """Add a new detection event"""
        event_data["timestamp"] = datetime.now()
        result = self.detection_events.insert_one(event_data)
        return str(result.inserted_id)
    
    def get_detection_events(self, camera_id: str, start_time: datetime, end_time: datetime) -> List[Dict]:
        """Get detection events for a camera within a time range"""
        return list(self.detection_events.find({
            "cameraId": camera_id,
            "timestamp": {
                "$gte": start_time,
                "$lte": end_time
            }
        }).sort("timestamp", -1))
    
    # Detection zones operations
    def configure_zones(self, camera_id: str, zones: List[Dict]) -> bool:
        """Configure detection zones for a camera"""
        result = self.detection_zones.update_one(
            {"cameraId": camera_id},
            {"$set": {"zones": zones}},
            upsert=True
        )
        return result.modified_count > 0 or result.upserted_id is not None
    
    def get_zones(self, camera_id: str) -> List[Dict]:
        """Get detection zones for a camera"""
        zones_doc = self.detection_zones.find_one({"cameraId": camera_id})
        return zones_doc.get("zones", []) if zones_doc else []
    
    # Device information operations
    def register_device(self, device_info: Dict) -> str:
        """Register a new device or update existing device info"""
        device_id = device_info.get("deviceId")
        if device_id:
            result = self.device_info.update_one(
                {"deviceId": device_id},
                {"$set": {**device_info, "lastSeen": datetime.now()}},
                upsert=True
            )
            return device_id
        else:
            device_info["registeredAt"] = datetime.now()
            device_info["lastSeen"] = datetime.now()
            result = self.device_info.insert_one(device_info)
            return str(result.inserted_id)
    
    def update_device_status(self, device_id: str, status: Dict) -> bool:
        """Update device status"""
        result = self.device_info.update_one(
            {"deviceId": device_id},
            {"$set": {
                "status": status,
                "lastSeen": datetime.now()
            }}
        )
        return result.modified_count > 0
    
    def get_device_info(self, device_id: str) -> Optional[Dict]:
        """Get device information"""
        return self.device_info.find_one({"deviceId": device_id})
    
    # Application state operations
    def save_app_state(self, user_id: str, state_data: Dict) -> bool:
        """Save application state for a user"""
        result = self.app_state.update_one(
            {"userId": user_id},
            {"$set": {
                "state": state_data,
                "updatedAt": datetime.now()
            }},
            upsert=True
        )
        return result.modified_count > 0 or result.upserted_id is not None
    
    def get_app_state(self, user_id: str) -> Optional[Dict]:
        """Get application state for a user"""
        state_doc = self.app_state.find_one({"userId": user_id})
        return state_doc.get("state", {}) if state_doc else None
    
    def close(self):
        """Close the MongoDB connection"""
        self.client.close()

File: frontend/src/lib/services/mongoDBService.ts
import { MongoClient, ServerApiVersion, ObjectId } from 'mongodb';
import { Camera, DetectionEvent, DetectionZone, DeviceInfo, AppState } from '@/lib/types';

// MongoDB Atlas connection string
const uri = process.env.NEXT_PUBLIC_MONGODB_URI || "mongodb+srv://testo:testo@storagecam.pgkdn.mongodb.net/?appName=storagecam";
const dbName = process.env.NEXT_PUBLIC_MONGODB_DATABASE || "storagecam";

// Create a MongoClient with a MongoClientOptions object to set the Stable API version
const client = new MongoClient(uri, {
  serverApi: {
    version: ServerApiVersion.v1,
    strict: true,
    deprecationErrors: true,
  }
});

// Collections
const collections = {
  cameras: () => client.db(dbName).collection('cameras'),
  detection_events: () => client.db(dbName).collection('detection_events'),
  detection_zones: () => client.db(dbName).collection('detection_zones'),
  device_info: () => client.db(dbName).collection('device_info'),
  app_state: () => client.db(dbName).collection('app_state')
};

export const mongoDBService = {
  // Connection management
  async connect(): Promise<void> {
    try {
      await client.connect();
      console.log("Connected to MongoDB Atlas");
    } catch (error) {
      console.error("Failed to connect to MongoDB:", error);
      throw error;
    }
  },

  async disconnect(): Promise<void> {
    await client.close();
    console.log("Disconnected from MongoDB Atlas");
  },

  async testConnection(): Promise<boolean> {
    try {
      await client.db("admin").command({ ping: 1 });
      console.log("Pinged your deployment. You successfully connected to MongoDB!");
      return true;
    } catch (error) {
      console.error("MongoDB connection test failed:", error);
      return false;
    }
  },

  // Camera operations
  async getCameras(userId: string): Promise<Camera[]> {
    try {
      const camerasCollection = collections.cameras();
      const result = await camerasCollection.find({ userId }).toArray();
      return result.map(doc => ({
        ...doc,
        id: doc._id.toString()
      })) as Camera[];
    } catch (error) {
      console.error("Error getting cameras:", error);
      throw error;
    }
  },
  
  async addCamera(data: any, userId: string): Promise<Camera> {
    try {
      const camerasCollection = collections.cameras();
      const cameraData = {
        ...data,
        userId,
        createdAt: new Date().getTime()
      };
      
      const result = await camerasCollection.insertOne(cameraData);
      return {
        ...cameraData,
        id: result.insertedId.toString()
      } as Camera;
    } catch (error) {
      console.error("Error adding camera:", error);
      throw error;
    }
  },
  
  async updateCamera(id: string, data: Partial<Camera>): Promise<Camera> {
    try {
      const camerasCollection = collections.cameras();
      await camerasCollection.updateOne(
        { _id: new ObjectId(id) },
        { $set: data }
      );
      
      const updatedCamera = await camerasCollection.findOne({ _id: new ObjectId(id) });
      return {
        ...updatedCamera,
        id: updatedCamera._id.toString()
      } as Camera;
    } catch (error) {
      console.error("Error updating camera:", error);
      throw error;
    }
  },
  
  async deleteCamera(id: string): Promise<void> {
    try {
      const camerasCollection = collections.cameras();
      await camerasCollection.deleteOne({ _id: new ObjectId(id) });
    } catch (error) {
      console.error("Error deleting camera:", error);
      throw error;
    }
  },
  
  // Detection events operations
  async getEvents(cameraId: string, startTime: number, endTime: number): Promise<DetectionEvent[]> {
    try {
      const eventsCollection = collections.detection_events();
      const result = await eventsCollection.find({
        cameraId,
        timestamp: {
          $gte: startTime,
          $lte: endTime
        }
      }).sort({ timestamp: -1 }).toArray();
      
      return result.map(doc => ({
        ...doc,
        id: doc._id.toString()
      })) as DetectionEvent[];
    } catch (error) {
      console.error("Error getting events:", error);
      throw error;
    }
  },
  
  async addEvent(eventData: Omit<DetectionEvent, 'id'>): Promise<string> {
    try {
      const eventsCollection = collections.detection_events();
      const result = await eventsCollection.insertOne({
        ...eventData,
        timestamp: eventData.timestamp || new Date().getTime()
      });
      return result.insertedId.toString();
    } catch (error) {
      console.error("Error adding event:", error);
      throw error;
    }
  },
  
  // Detection zones operations
  async configureZones(cameraId: string, zones: DetectionZone[]): Promise<void> {
    try {
      const zonesCollection = collections.detection_zones();
      await zonesCollection.updateOne(
        { cameraId },
        { $set: { zones } },
        { upsert: true }
      );
    } catch (error) {
      console.error("Error configuring zones:", error);
      throw error;
    }
  },
  
  async getZones(cameraId: string): Promise<DetectionZone[]> {
    try {
      const zonesCollection = collections.detection_zones();
      const result = await zonesCollection.findOne({ cameraId });
      return result?.zones || [];
    } catch (error) {
      console.error("Error getting zones:", error);
      throw error;
    }
  },
  async registerDevice(deviceInfo: DeviceInfo): Promise<string> {
    try {
      const deviceCollection = collections.device_info();
      const now = new Date().getTime();
      
      if (deviceInfo.deviceId) {
        // Update existing device
        await deviceCollection.updateOne(
          { deviceId: deviceInfo.deviceId },
          { 
            $set: {
              ...deviceInfo,
              lastSeen: now
            }
          },
          { upsert: true }
        );
        return deviceInfo.deviceId;
      } else {
        // Create new device
        const newDevice = {
          ...deviceInfo,
          registeredAt: now,
          lastSeen: now
        };
        const result = await deviceCollection.insertOne(newDevice);
        return result.insertedId.toString();
      }
    } catch (error) {
      console.error("Error registering device:", error);
      throw error;
    }
  }
  
  async updateDeviceStatus(deviceId: string, status: any): Promise<void> {
    try {
      const deviceCollection = collections.device_info();
      await deviceCollection.updateOne(
        { deviceId },
        {
          $set: {
            status,
            lastSeen: new Date().getTime()
          }
        }
      );
    } catch (error) {
      console.error("Error updating device status:", error);
      throw error;
    }
  },
  
  async getDeviceInfo(deviceId: string): Promise<DeviceInfo> {
    try {
      const deviceCollection = collections.device_info();
      const result = await deviceCollection.findOne({ deviceId });
      return result as DeviceInfo;
    } catch (error) {
      console.error("Error getting device info:", error);
      throw error;
    }
  },
  
  // Application state operations
  async saveAppState(userId: string, state: AppState): Promise<void> {
    try {
      const stateCollection = collections.app_state();
      await stateCollection.updateOne(
        { userId },
        {
          $set: {
            state,
            updatedAt: new Date().getTime()
          }
        },
        { upsert: true }
      );
    } catch (error) {
      console.error("Error saving app state:", error);
      throw error;
    }
  },
  
  async getAppState(userId: string): Promise<AppState> {
    try {
      const stateCollection = collections.app_state();
      const result = await stateCollection.findOne({ userId });
      return result?.state as AppState || null;
    } catch (error) {
      console.error("Error getting app state:", error);
      throw error;
    }
  }
};

// Initialize connection when the service is imported
mongoDBService.connect().catch(console.error);
